### PR TITLE
generator; single main() entrypoint for each suite.

### PR DESCRIPTION
**Pull request by raulk**
This introduces a new organisation for test vector scripts.

Each suite now lives in a package (directory), with a single main entrypoint.

I have introduced a `Generator` that the main method can delegate to to have it handle the generation and writing of test vectors into properly named files. I have also taken the opportunity to make things DRYer; scripts no longer have to instantiate the `Builder` manually, or call `v.Finish()`: the `Generator` takes care of that.

See godocs of the `Generator` struct for more info on the layout.

I have migrated the `message_application` suite to use this style. `paych` suite pending.

---

This fixes https://github.com/filecoin-project/oni/issues/189 via an alternative approach that is less hacky and a lot more declarative, and DRY.

---

Try it out:

```shell script
$ cd tvx
$ go run ./scripts/msg_application -o ./foo
generating test vector: msg-apply-fail-actor-execution-illegal-arg
generating test vector: msg-apply-fail-unknown-receiver
generating test vector: msg-apply-fail-unknown-receiver
generating test vector: msg-apply-fail-invalid-receiver-method
generating test vector: msg-apply-fail-invalid-receiver-method
generating test vector: msg-apply-fail-onchainsize-gas
generating test vector: msg-apply-fail-transfer-accountcreation-gas
generating test vector: msg-apply-fail-transfer-accountcreation-gas
$ ls -lrt foo
total 208
-rw-r--r--  1 raul  staff   2720 Aug 14 15:59 invalid_msgs-0.json
-rw-r--r--  1 raul  staff   2720 Aug 14 15:59 invalid_msgs-1.json
-rw-r--r--  1 raul  staff   3142 Aug 14 15:59 unknown_actors-1.json
-rw-r--r--  1 raul  staff   3142 Aug 14 15:59 unknown_actors-0.json
-rw-r--r--  1 raul  staff   4045 Aug 14 15:59 actor_exec-0.json
-rw-r--r--  1 raul  staff  26436 Aug 14 15:59 gas_cost-2.json
-rw-r--r--  1 raul  staff  26436 Aug 14 15:59 gas_cost-1.json
-rw-r--r--  1 raul  staff  26436 Aug 14 15:59 gas_cost-0.json
```

### CODE REVIEW

File: tvx/builders/generator.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	)
+
+	flag.Parse()
+
+	ret := new(Generator)
+
+	// Output directory is compulsory.
+	if *outputDir == "" {
+		fmt.Printf("output directory not specified\n")
+		os.Exit(1)
```
Comment (nonsense) [inline tvx/builders/generator.go:62]:
```suggestion
		log.Fatalf("output directory not specified\n")
```

File: tvx/builders/generator.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+	}
+
+	err := ensureDirectory(*outputDir)
+	if err != nil {
+		log.Println(err)
+		os.Exit(1)
```
Comment (nonsense) [inline tvx/builders/generator.go:68]:
```suggestion
		log.Fatal(err)
```

File: tvx/builders/generator.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+	}
+
+	ret.OutputPath = *outputDir
+
+	// If a filter has been provided, compile it into a regex.
+	if *filter != "" {
+		exp, err := regexp.Compile(*filter)
+		if err != nil {
+			fmt.Printf("supplied regex %s is invalid: %s\n", *filter, err)
+			os.Exit(1)
```
Comment (nonsense) [inline tvx/builders/generator.go:78]:
```suggestion
			log.Fatalf("supplied regex %s is invalid: %s\n", *filter, err)
```

File: tvx/builders/generator.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+			out, err := os.OpenFile(file, os.O_WRONLY|os.O_CREATE, 0644)
+			if err != nil {
+				fmt.Printf("failed to write to file %s: %s\n", file, err)
+				return
+			}
+			wg.Add(1)
+			go func(item *MessageVectorGenItem) {
+				g.generateOne(out, item)
+				wg.Done()
+			}(&item)
```
Comment (nonsense) [inline tvx/builders/generator.go:112]: Is that right? It is not obvious what the type of `item` is, so maybe just `item := item` to scope it?

File: tvx/builders/generator.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+//  -o <directory>
+//		directory where test vector JSON files will be saved.
+//
+//  -f <regex>
+//		regex filter to select a subset of vectors to execute; matched against
+//	 	the vector's ID.
+//
+// Scripts can bundle test vectors into "groups". The generator will execute
+// each group in parallel, and will write each vector in a file:
+// <output_dir>/<group>-<sequential>.json
```
Comment (alanshaw) [inline tvx/builders/generator.go:35]:
```suggestion
// <output_dir>/<group>-<vector_ID>.json
```

Might be more stable and not cause problems when re-generating vectors over existing files (easier to identify also).

File: tvx/builders/generator.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+		outputDir = flag.String("o", "", "output directory")
+		filter    = flag.String("f", "", "filter regex for test vectors to generate")
+	)
+
+	flag.Parse()
+
+	ret := new(Generator)
+
+	// Output directory is compulsory.
+	if *outputDir == "" {
```
Comment (alanshaw) [inline tvx/builders/generator.go:60]: Hmm, I thought the idea was to output to `os.Stdout` if this was not set?

Comment (raulk) [reply tvx/builders/generator.go:60]: Yup! I'm gonna implement that now. Was thinking on the spot during the call ;-)

Comment (alanshaw) [reply tvx/builders/generator.go:60]: I like your style 😁

File: tvx/builders/generator.go (hunk 5 above)
Comment (raulk) [reply tvx/builders/generator.go:35]: Done!

File: tvx/builders/generator.go (hunk 3 above)
Comment (raulk) [reply tvx/builders/generator.go:78]: Implemented.

File: tvx/builders/generator.go (hunk 2 above)
Comment (raulk) [reply tvx/builders/generator.go:68]: Implemented.

File: tvx/builders/generator.go (hunk 1 above)
Comment (raulk) [reply tvx/builders/generator.go:62]: Implemented.

Context: we were suppressing logs in the `init` function of `builder.go`, to not muddy the output JSON, that's why I was using `fmt.Print*` directly. But @willscott suggested to send the log package to `os.Stderr`, which keeps Unix pipes working, but it muddies the console output. To make things better, I have prefix `log` output with `>>>`.

File: tvx/builders/generator.go (hunk 4 above)
Comment (raulk) [reply tvx/builders/generator.go:112]: Fixed.

**Pull request merged**